ci/Jenkinsfile.test-e2e.android

```diff
@@ -0,0 +1,277 @@
+                  -F "file=@${env.APK_PATH}" \\
+                  -F "custom_id=\${CUSTOM_ID}" \\
+                | tee upload_response.json
+              """,
+              returnStdout: true
+            ).trim()
+            def result = readJSON text: response
+            def appUrl = result?.app_url
+            if (!appUrl) {
+              error("BrowserStack upload failed: ${response}")
+            }
+            env.BROWSERSTACK_APP_ID = appUrl
+            env.BROWSERSTACK_BUILD_NAME = result?.custom_id
+            env.BROWSERSTACK_BUILD_IDENTIFIER = env.BUILD_NUMBER
+            echo "BrowserStack app uploaded: ${env.BROWSERSTACK_APP_ID}"
+          }
+        }
+      }
+    }
+
+    stage('Run pytest suite') {
+      steps {
+        script {
+          dir('test/e2e_appium') {
+            if (!env.DEVICE_ID?.trim()) {
+              echo 'DEVICE_ID not set; using framework defaults.'
+            } else {
+              echo "Using DEVICE_ID: ${env.DEVICE_ID}"
+            }
+
+            def pytestCmd = "python -m pytest --env browserstack"
+            if (env.PYTEST_ARGS?.trim()) {
+              pytestCmd += " ${env.PYTEST_ARGS.trim()}"
+            }
+            
+            withCredentials([
+              usernamePassword(
+                credentialsId: 'browserstack-status-desktop',
+                usernameVariable: 'BROWSERSTACK_USERNAME',
+                passwordVariable: 'BROWSERSTACK_ACCESS_KEY'
+              )
+            ]) {
+              withEnv([
+                "BROWSERSTACK_USERNAME=${BROWSERSTACK_USERNAME}",
+                "BROWSERSTACK_ACCESS_KEY=${BROWSERSTACK_ACCESS_KEY}",
+                "BROWSERSTACK_APP_ID=${env.BROWSERSTACK_APP_ID}",
+                "BROWSERSTACK_BUILD_NAME=${env.BROWSERSTACK_BUILD_NAME ?: ''}",
+                "BROWSERSTACK_BUILD_IDENTIFIER=${env.BROWSERSTACK_BUILD_IDENTIFIER ?: ''}",
+                "BROWSERSTACK_PROJECT_NAME=${env.BROWSERSTACK_PROJECT_NAME ?: ''}",
+                "TEST_DEVICE_ID=${env.DEVICE_ID}"
+              ]) {
+                sh """
+                  set -euo pipefail
+                  source ${VIRTUAL_ENV}/bin/activate
+                  ${pytestCmd}
+                """
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  post {
+    success {
+      script {
+        github.notifyPR(true)
+      }
+    }
+    failure {
+      script {
+        github.notifyPR(false)
+      }
+    }
+    always {
+      script {
+        def runId = env.E2E_RUN_ID?.trim()
+        def reportsPattern = runId ? "test/e2e_appium/reports/${runId}/**/*.xml" : "test/e2e_appium/reports/**/*.xml"
+        junit allowEmptyResults: true, testResults: reportsPattern
+        def archivePattern = runId ? "test/e2e_appium/reports/${runId}/**/*" : "test/e2e_appium/reports/**/*"
+        archiveArtifacts artifacts: archivePattern, allowEmptyArchive: true
```

Comment:
I would recommend not putting things like `junit` or `archiveArtifacts` in `always` block. It should be its own stage. Makes it easier to identify and debug issues with specific stages of the job.